clients/drcachesim/scheduler/scheduler_impl.h

```diff
@@ -195,8 +195,27 @@ template <typename RecordType, typename ReaderType> class scheduler_impl_tmpl_t
         // If non-empty these records should be returned before incrementing the reader.
         // This is used for read-ahead and inserting synthetic records.
         // We use a deque so we can iterate over it.
-        std::deque<RecordType> queue;
+        struct queued_record_t {
+            queued_record_t()
+            {
+            }
+            queued_record_t(const RecordType &record)
+                : record(record)
+            {
+            }
+            queued_record_t(const RecordType &record, uint64_t next_trace_pc)
+                : record(record)
+                , next_trace_pc(next_trace_pc)
+            {
+            }
+
+            RecordType record;
+            uint64_t next_trace_pc = 0;
+        };
+        std::deque<queued_record_t> queue;
         bool cur_from_queue;
+        queued_record_t cur_queue_record;
```

Comment:
Add a comment that there is an invariant that if the front of the queue is to be returned, it has to be removed from the queue and stored here and cur_from_queue set (presumably cur_from_queue is then cleared after returning it).

Is it better to *not* remove it from the queue until it's returned? Can we then get rid of cur_from_queue *and* cur_queue_record? That seems simpler.

Reply:
If we let the record from the queue be on the queue's front until the next iteration, then we need cur_from_queue to remind us to remove the queue front. Keeping the record on the queue also would make logic that adds to the front of the queue more complicated.

Reply:
> We could remove from the queue right when we return it: no need to wait until the next next_record call.

That's what we're doing now too. We need to keep `cur_queue_record` because it may hold information required to answer the stream queries (the next trace pc in this case).